lib/connectors/mongodb/connector.rb

```diff
@@ -57,10 +59,23 @@ def initialize(configuration: {})
 
       def yield_documents
         with_client do |client|
-          client[@collection].find.each do |document|
-            doc = document.with_indifferent_access
-
-            yield serialize(doc)
+          # We do paging using skip().limit() here to make Ruby recycle the memory for each page pulled from the server after it's not needed any more.
+          # This gives us more control on the usage of the memory (we can adjust PAGE_SIZE constant for that to decrease max memory consumption).
+          # It's done due to the fact that usage of .find.each leads to memory leaks or overuse of memory - the whole result set seems to stay in memory
+          # during the sync. Sometimes (not 100% sure) it even leads to a real leak, when the memory for these objects is never recycled.
+          cursor = client[@collection].find
+          skip = 0
+
+          loop do
+            found_count = 0
+            view = cursor.skip(skip).limit(PAGE_SIZE)
+            view.each do |document|
+              yield serialize(document)
+              found_count += 1
+            end
+
+            break if found_count == 0
+            skip += PAGE_SIZE
           end
```

Comment:
```suggestion
          skip = 0

          loop do
            view = client[@collection].find.skip(skip).limit(PAGE_SIZE)
            break if view.count == 0
            view.each do |document|
              yield serialize(document)
            end
            skip += PAGE_SIZE
          end
```

I didn't do any testing, but it seems `client[@collection].find` just [returns a View](https://github.com/mongodb/mongo-ruby-driver/blob/15a77ccf0dc5c6d1fbd49754358a95b019664190/lib/mongo/collection.rb#L502), which [include Immutable](https://github.com/mongodb/mongo-ruby-driver/blob/15a77ccf0dc5c6d1fbd49754358a95b019664190/lib/mongo/collection/view.rb#L49). So when you call `skip` or `limit`, it will just [return a new View](https://github.com/mongodb/mongo-ruby-driver/blob/15a77ccf0dc5c6d1fbd49754358a95b019664190/lib/mongo/collection/view/immutable.rb#L34). I don't see any point to assign it to `cursor` here.

Reply:
Tested it, indeed everything is as you described, thanks for checking!